// configure/configure.js
//sqlserver驱动配置
var driver = 'SQL Server Native Client 10.0';
var server = '(local)';
var user = '<user>';
var pwd = '<pwd>';
var database = 'test';
var useTrustedConnection = true;
var sqlserver = "Driver={" + driver + "};Server=" + server + ";" + (useTrustedConnection == true ? "Trusted_Connection={Yes};" : "UID=" + user + ";PWD=" + pwd + ";") + "Database={" + database + "};";

var pSpace ={
		"host":"localhost:8889",
		"user":"admin",
		"password":"admin888"
	};
var workPath = "D:/his64";
var interVal =2000;
var configPath = "D:/his64/config.csv";
var isSaveToSql =true;

// The following need to be exported for building connection strings within a test...
exports.database = database;
exports.server = server;
exports.user = user;
exports.pwd = pwd;
// Driver name needs to be exported for building expected error messages...
exports.driver = driver;
// Here's a complete connection string which can be shared by multiple tests...
exports.sqlserver = sqlserver;
exports.pSpace = pSpace;
exports.workPath = workPath;
exports.interVal = interVal;
exports.configPath = configPath;
exports.isSaveToSql = isSaveToSql;
